### PR TITLE
Add rest health checker

### PR DESCRIPTION
**Pull request by Y-Sindo**
In Mgm SDK, under persistent mode each negotiation will check the endpoint health via the status of WebSocket connection. However transient mode doesn't have this mechanism. So we need a health checker to check the endpoint health via REST api periodically.

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 1]
```diff
@@ -1,22 +1,84 @@
...
+            _logger = logger;
+            _hubName = hubName;
+            _cancellationTokenSource = new();
+            _cancellationToken = _cancellationTokenSource.Token;
+        }
+
         public Task StartAsync(CancellationToken cancellationToken)
         {
-            return Task.CompletedTask;
+            _timer = new Timer(CheckEndpointHealthAsync, null, TimeSpan.Zero, CheckInterval);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:41]: try `TimeAwaitable`

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 2]
```diff
@@ -1,22 +1,84 @@
...
+        public Task StopAsync(CancellationToken _)
         {
+            _cancellationTokenSource.Cancel();
+            _firstCheckCompletionSource.TrySetCanceled();
             return Task.CompletedTask;
         }
+
+        private async void CheckEndpointHealthAsync(object _)
+        {
+            var endpoints = _serviceEndpointManager.GetEndpoints(_hubName);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:54]: not used?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 3]
```diff
@@ -1,22 +1,84 @@
...
+                    var isHealthy = await client.IsServiceHealthy(_cancellationToken);
+                    endpoint.Online = isHealthy;
+                }
+                catch (OperationCanceledException ex) when (ex.CancellationToken == _cancellationToken)
+                {
+                    // the health checker is stopping.
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Check health status failed for endpoint {endpoint}", endpoint.Endpoint);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:69]: use the standard log fomat

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:69]: warning is enough?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 4]
```diff
@@ -1,22 +1,84 @@
...
+                    endpoint.Online = false;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:70]: is this too strict that when health check throws return false?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 5]
```diff
@@ -1,22 +1,84 @@
...
+                }
+                finally
+                {
+                    _firstCheckCompletionSource.TrySetResult(true);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:74]: when one endpoint finishes healthy check, return true? what if error throws?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 6]
```diff
@@ -1,22 +1,84 @@
...
 
-        public Task StopAsync(CancellationToken cancellationToken)
+        public Task StopAsync(CancellationToken _)
         {
+            _cancellationTokenSource.Cancel();
+            _firstCheckCompletionSource.TrySetCanceled();
             return Task.CompletedTask;
         }
+
+        private async void CheckEndpointHealthAsync(object _)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:52]: intentionally use `void`?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs (hunk 1 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:41]: What is the usage of `TimeAwaitable`?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs (hunk 4 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:60]: what if client http timeout?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 7]
```diff
@@ -1,22 +1,110 @@
...
+                var isHealthy = false;
+                bool needRetry;
+                do
+                {
+                    try
+                    {
+                        using var client = _clientFactory.Create(endpoint);
+                        isHealthy = await client.IsServiceHealthy(default);
+                    }
+                    catch (Exception ex)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:63]: when it throws `isHealthy` is not set to `false`, which means it is possible that isHealthy keeps last `true` value, is it by design?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs [hunk 8]
```diff
@@ -1,22 +1,110 @@
...
+                while (await _timer)
+                {
+                    await CheckEndpointHealthAsync();
+                }
+            }
+        }
+
+        private static class Log
+        {
+            private static readonly Action<ILogger, string, Exception> _restHealthCheckFailed = LoggerMessage.Define<string>(LogLevel.Information, new EventId(1, nameof(RestHealthCheckFailed)), "Failed to check health state for endpoint {endpoint}");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:93]: make the error message more like `Information` log

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs (hunk 7 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs:63]: The local var `isHealthy` is initialized to `false`. If exception is thrown, `isHealthy` is not set new value and won't keep last true value.

**Review by vicancy — APPROVED** [review]

**Pull request merged**